horizon/facts/router.py

```diff
@@ -47,73 +73,199 @@ async def replace_user(
     request: FastApiRequest,
     client: FactsClientDependency,
     update_subscriber: DataUpdateSubscriberDependency,
+    wait_timeout: WaitTimeoutDependency,
     user_id: str,
 ):
     return await forward_request_then_wait_for_update(
         client,
         request,
         update_subscriber,
+        wait_timeout,
         path=f"/users/{user_id}",
-        obj_type="user",
+        entries_callback=lambda r, body: [
+            create_data_source_entry(
+                obj_type="users",
+                obj_id=body["id"],
+                obj_key=body["key"],
+                authorization_header=r.headers.get("Authorization"),
+            )
+        ],
+    )
+
+
+def create_role_assignment_data_entries(
+    request: FastApiRequest, body: dict[str, Any]
+) -> Iterable[DataSourceEntry]:
+    if not body.get("resource_instance"):
+        yield create_data_source_entry(
+            obj_type="role_assignments",
+            obj_id=body["id"],
+            obj_key=f"user:{body['user']}",
+            authorization_header=request.headers.get("Authorization"),
+        )
+        yield create_data_source_entry(
+            obj_type="users",
+            obj_id=body["user_id"],
+            obj_key=body["user"],
+            authorization_header=request.headers.get("Authorization"),
+        )
+    else:
+        # note that user_id == subject_id,
+        # and user == user_key == subject_key == subject_str
+        yield create_data_source_entry(
+            obj_type="role_assignments",
+            obj_id=body["user_id"],
+            obj_key=body["user"],
+            authorization_header=request.headers.get("Authorization"),
+        )
+
+
+@facts_router.post("/users/{user_id}/roles")
+async def assign_user_role(
+    request: FastApiRequest,
+    client: FactsClientDependency,
+    update_subscriber: DataUpdateSubscriberDependency,
+    wait_timeout: WaitTimeoutDependency,
+    user_id: str,
+):
+    return await forward_request_then_wait_for_update(
+        client,
+        request,
+        update_subscriber,
+        wait_timeout,
+        path=f"/users/{user_id}/roles",
```

Comment:
I think you can use request.path or something similar ( with removing the /facts prefix instead of "copy pasting" the path from the route declaration in every function. 
Isn't a major thing